op-chain-ops: implement withdrawal hashing

tynes opened the pull request:
**Description**

Implement both the new hashing scheme and the legacy hashing for withdrawals so that the withdrawals can be migrated.
This only includes tests for legacy style hashing, will follow up with an additional PR to add tests for bedrock style withdrawal hashing to prevent this PR from growing too large.

Closes ENG-2718

<!--
Please fill in each sections of this template, and delete any sections that are not relevant.

Need help?
Refer to our contributing guidelines for additional information about making a good pull request:
https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md
-->

tynes commented:
Can get test vectors from here: https://optimistic.etherscan.io/address/0x4200000000000000000000000000000000000000#internaltx

tynes commented on op-chain-ops/crossdomain/withdrawal.go:
```diff
@@ -0,0 +1,112 @@
+	return enc, nil
+}
+
+// EncodeLegacy will serialze the Withdrawal in the legacy format so that it
```
Perhaps this should be `LegacyEncode` to be the same as the other legacy functions (where legacy is the prefix) or the others should be updated such that legacy is the suffix

trianglesphere reviewed (commented):
LGTM - one Q about some test setup. I really like how it compares the computed slot vs what get set in the trace.

mslipper commented:
Recusing myself from review here; others will have more insightful feedback.

tynes commented:
> LGTM - one Q about some test setup. I really like how it compares the computed slot vs what get set in the trace.

Good call on catching that error, I have updated it to panic on error of abi deserialization

tynes commented:
I've added some additional functionality to this PR:
- Split the withdrawal type into `LegacyWithdrawal` and `Withdrawal`
- encoding + decoding with roundtrip fuzz tests of both withdrawal types
- Test coverage of the storage slot generation for the `Withdrawal` type
  - Test vectors generated using foundry, this is important for the storage slot migration

smartcontracts commented on op-chain-ops/crossdomain/message.go:
```diff
@@ -67,6 +68,26 @@ func (c *CrossDomainMessage) Hash() (common.Hash, error) {
+	}
+}
+
+// ToWithdrawal will turn a CrossDomainMessage into a Withdrawal.
```
I'm still a little unclear about what this function is supposed to do

tynes replied:
V0 cross domain messages have all of the information required to create the corresponding withdrawal type, so this is a helper to get access to the the methods on the withdrawal given the same data in the cross domain message

smartcontracts replied:
Oh ok I see, thanks for the clarification!

trianglesphere commented on op-chain-ops/crossdomain/legacy_withdrawal.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+// Decode will decode a serialized LegacyWithdrawal
+func (w *LegacyWithdrawal) Decode(data []byte) error {
```
I'm not sure how critical this is given that this isn't exposed to much hostile input, but this will panic if `data` is not long enough.

tynes replied:
Ah good catch, might as well add checks for that

trianglesphere commented on op-chain-ops/crossdomain/withdrawal.go:
```diff
@@ -0,0 +1,116 @@
+	}
+
+	sender, ok := decoded[1].(common.Address)
+	if !ok {
```
Why are sender/target done with the `ok` syntax (no panics), but the rest are done a in a manner that could panic (if the type was not correct in the result of `args.Unpack`).

trianglesphere replied:
Ok. I agree that this is very unlikely to cause a problem, I was curious why there was a difference.

tynes commented:
> 

i have addressed the issues, thanks for the review

trianglesphere commented on op-chain-ops/crossdomain/legacy_withdrawal.go:
```diff
@@ -0,0 +1,132 @@
+
+// Decode will decode a serialized LegacyWithdrawal
+func (w *LegacyWithdrawal) Decode(data []byte) error {
+	if len(data) < 4 {
```
The length check should be that is greater than 4 + len(address). This is because on line 71 we slice from `4` to `len(data) - len(address)`. We have only guaranteed that `len(data) >= len(address) >= 4`. That means that if `len(data) == len(address)`, slicing from `4:0` on line 71 would panic.

tynes replied:
I've addressed this in https://github.com/ethereum-optimism/optimism/commit/dd274000bcbdc3325f11e84918d6381ced10786c

trianglesphere reviewed (changes requested):
There's a small bug in the length check code. Fix that and it's good to go.

tynes commented:
Not sure why this is failing in CI but not locally for me

mslipper commented:
It fails for me locally as well.

trianglesphere commented:
@tynes is there an issue with the files checked in / something around the path where it's not able to load in the test cases?

tynes commented:
> @tynes is there an issue with the files checked in / something around the path where it's not able to load in the test cases?

Found the source of the bug, it was due to the way that the L2ToL1MessagePasser is now being handled. Instead of migrating the state at `address(0x42..00)` we are introducing a new predeploy at `address(0x42..16)` that will receive the migrated hashes and the proof on L1 will work against that account.
